orttraining/orttraining/core/optimizer/gist_encode_decode.cc

```diff
@@ -45,74 +45,115 @@ static std::vector<GraphEdgeHelper> GetNodeOutputEdges(const Node& node) {
+    ONNX_NAMESPACE::TypeProto compressed_tensor;
+    compression_type = user_compression_type;
 
-  bool early_encoding = false;
-  bool late_decoding = false;
-  for (auto& output_edge : output_edges) {
-    Node* node_dst = graph.GetNode(output_edge.dst_node);
-    if (node_dst->Description() == "Backward pass" && (node_dst->OpType() == "ReluGrad")) {
-      graph.AddEdge(output_edge.src_node, encode.Index(), output_edge.src_arg_index, 0);
-      graph.AddEdge(encode.Index(), decode.Index(), 1, 1);
-      graph.AddEdge(decode.Index(), output_edge.dst_node, 0, output_edge.dst_arg_index);
-      std::vector<GraphEdgeHelper> input_edges_dst = GetNodeInputEdges(*node_dst);
-      size_t i = 0;
-      while (!late_decoding && i < input_edges_dst.size()) {
-        if (graph.GetNode(input_edges_dst[i].src_node)->OpType() != curr_node.OpType()) {
-          graph.AddEdge(input_edges_dst[i].src_node, decode.Index(), input_edges_dst[i].src_arg_index, 0);
-          late_decoding = true;
+    // Override compression_type for lossless compression case(s) (eg. bool -> Pack1)
+    ONNX_NAMESPACE::DataType type_string = curr_node_output_arg->Type();
+    if (*type_string == "bool" || *type_string == "tensor(bool)") {
+      std::cout << "(Lossless) override compression type to Pack1 for tensor: " << curr_node_output_arg->Name() << std::endl;
+      compression_type = "GistPack1";
+    }
+    
+    if (compression_type == "GistPack1" || compression_type == "GistPack8" || compression_type == "GistPackMsfp15") {
+      compressed_tensor.mutable_tensor_type()->set_elem_type(ONNX_NAMESPACE::TensorProto_DataType_UINT8);
+    }
+    else if (compression_type == "GistPack16") {
+      compressed_tensor.mutable_tensor_type()->set_elem_type(ONNX_NAMESPACE::TensorProto_DataType_FLOAT16);
+    }
+    else {
+      assert (0); // "Gist compression type not supported"
+    }
+    
+    bool tensor_size_compressed = false;
+    for (int i = 0; i < curr_node_output_arg->Shape()->dim_size(); i++) {
+      if (curr_node_output_arg->Shape()->dim(i).dim_value() != 0) {
+        if (compression_type == "GistPack1" && !tensor_size_compressed) {
+          compressed_tensor.mutable_tensor_type()->mutable_shape()->add_dim()->set_dim_value((curr_node_output_arg->Shape()->dim(i).dim_value()  + GIST_PACK1_FACTOR -1)/GIST_PACK1_FACTOR); 
+          tensor_size_compressed = true;
         }
-        i++;
+        else {
+          compressed_tensor.mutable_tensor_type()->mutable_shape()->add_dim()->set_dim_value(curr_node_output_arg->Shape()->dim(i).dim_value());
+        } 
       }
-    } else if (!early_encoding) {
-      graph.AddEdge(encode.Index(), output_edge.dst_node, 0, output_edge.dst_arg_index);
-      early_encoding = true;
+      else {
+        compressed_tensor.mutable_tensor_type()->mutable_shape()->add_dim()->set_dim_param(curr_node_output_arg->Shape()->dim(i).dim_param());
+      }
+    }
```

Comment:
Can this part be handled by shape inference on the Gist ops?

Reply:
For Gist we dynamically reshape the compressed tensor and this might fall into the limitation of ShapeInference as is listed  [here](https://github.com/onnx/onnx/blob/master/docs/ShapeInference.md):
`Shape inference is not guaranteed to be complete. In particular, some dynamic behaviors block the flow of shape inference, for example a Reshape to a dynamically-provide shape. Also, all operators are not required to have a shape inference implementation.`

We would like to skip ShapeInference on this PR, and we could upgrade Gist with ShapeInference in a future PR.

Reply:
Fair enough to do this in a later PR, although I'm not sure how putting the code here avoids that fundamental limitation of static shape inference. Aren't you still limited to whatever information is available on the original graph, which might be incomplete?

Reply:
Yes, when setting the dimensions for the compressed tensor here, we are only using the information in the original graph. Based on our current understanding, this may be restrictive for certain compression cases.

It seems that the current implementation can be moved to Shape Inference logic. We will update this in a later PR.